### PR TITLE
create Academic Discipline endpoint

### PR DESCRIPTION
**Pull request by hieuppham**
In Discipline model, these files are ignored cuz they are unnecessary:
- types
- source
- source_url
/api/discipline has these functions:
- Find disciplines and sort by either name(english and vietnamese) or level
- Find discipline by id
- Search discipline by name and synonym (english and vietnamese)
- Find disciplines by parent_id
- Get the number of disciplines in each parent (using customRepo)
- Add new discipline
- Delete discipline
- Update discipline

### CODE REVIEW

**Comment by hieuppham** [comment]
In AcademicRank model, these fields are ignored cuz they are unecessary:
- types
- source
/api/academic-rank endpoint has these functions:
- find ranks and sort by name (not synonyms)
- find rank by id
- Search by name(english only) and synonym(english and vietnamese)
- Add new rank
- delete rank
- update rank

File: src/main/java/org/vietsearch/essme/controller/DisciplineController.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+        Page<Discipline> disciplinePage = disciplineRepository.findAll(PageRequest.of(page, size, sort));
+        return disciplinePage.getContent();
+    }
+
+    @GetMapping("/{_id}")
+    public Discipline getDisciplineById(@PathVariable("_id") String _id) {
+        Optional<Discipline> optionalDiscipline = disciplineRepository.findById(_id);
+        optionalDiscipline.orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Discipline not found"));
+        return optionalDiscipline.get();
```
Comment (btl01) [inline src/main/java/org/vietsearch/essme/controller/DisciplineController.java:52]:
```
return disciplineRepository.findById(_id).orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Discipline not found"));

**Comment by hieuppham** [comment]
In Company model, these fields are ignored cuz they are unecessary:
- types
- source
/api/company endpoint has these functions:
- find and sort companies by either name(in 4 languages) or rank (3 kinds of rank)
- find companies by both country name both industry, than sort by rank (3 kinds of rank)
- search by either name or synonym (in 4 languages)
- get by id
- add new company
- delete company
- update company

**Comment by btl01** [comment]
Bạn xem hàm search của mình nhé
```
    @GetMapping("/search")
    public List<Corporate> search(@RequestParam("name") String name) {
        TextCriteria criteria = TextCriteria.forDefaultLanguage().matchingPhrase(name);
        List<Corporate> list = corporateRepository.findBy(criteria);
        if (list.isEmpty())
            return corporateRepository.findByNameStartsWithIgnoreCase(name);
        return list;
    }
```
repository
```
public interface CorporateRepository extends MongoRepository<Corporate, String> {
    List<Corporate> findBy(TextCriteria criteria);
    List<Corporate> findByNameStartsWithIgnoreCase(String name);
    Corporate findByNameIgnoreCase(String name);
}
```
Gõ tiếng Anh gợi ý cực mạnh

**Comment by hieuppham** [comment]
> Bạn xem hàm search của mình nhé
> 
> ```
>     @GetMapping("/search")
>     public List<Corporate> search(@RequestParam("name") String name) {
>         TextCriteria criteria = TextCriteria.forDefaultLanguage().matchingPhrase(name);
>         List<Corporate> list = corporateRepository.findBy(criteria);
>         if (list.isEmpty())
>             return corporateRepository.findByNameStartsWithIgnoreCase(name);
>         return list;
>     }
> ```
> 
> repository
> 
> ```
> public interface CorporateRepository extends MongoRepository<Corporate, String> {
>     List<Corporate> findBy(TextCriteria criteria);
>     List<Corporate> findByNameStartsWithIgnoreCase(String name);
>     Corporate findByNameIgnoreCase(String name);
> }
> ```
> 
> Gõ tiếng Anh gợi ý cực mạnh

Ok bro, tại t không biết có cái này. Thanks

**Comment by btl01** [comment]
Không biết 'name' của cái này có unique không?
review code mình nhé https://github.com/Lampx83/essme-backend/commit/97d84539bb8f37f06ce6dcd6e3d12593ea64c880 corporate commit

**Comment by hieuppham** [comment]
> Không biết 'name' của cái này có unique không? review code mình nhé [97d8453](https://github.com/Lampx83/essme-backend/commit/97d84539bb8f37f06ce6dcd6e3d12593ea64c880) corporate commit

Oh great. Code vội để còn xem phim :)

**Comment by btl01** [comment]
Bao giờ sửa xong merge nha b

**Comment by btl01** [comment]
Bao giờ sửa xong merge nha b

**Comment by btl01** [comment]
Thêm xong merge nha b

**Pull request merged**